File: stream-processing/flink/examples/java/src/main/java/com/example/flink/KafkaToKafka.java
package com.example.flink;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.flink.connector.base.DeliveryGuarantee;
import org.apache.flink.connector.kafka.sink.KafkaRecordSerializationSchema;
import org.apache.flink.connector.kafka.sink.KafkaSink;
import org.apache.flink.connector.kafka.source.KafkaSource;
import org.apache.flink.connector.kafka.source.enumerator.initializer.OffsetsInitializer;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

/**
 * Kafka to Kafka Pipeline
 * 
 * Complete pipeline that:
 * 1. Reads from Kafka topic (source-topic)
 * 2. Transforms the data
 * 3. Writes to Kafka topic (sink-topic)
 * 
 * Prerequisites:
 * - Kafka must be running
 * - Topics 'source-topic' and 'sink-topic' must exist
 * 
 * Create topics:
 * docker exec -it kafka kafka-topics --create \
 *   --topic source-topic --bootstrap-server localhost:9092 \
 *   --partitions 3 --replication-factor 1
 * 
 * docker exec -it kafka kafka-topics --create \
 *   --topic sink-topic --bootstrap-server localhost:9092 \
 *   --partitions 3 --replication-factor 1
 * 
 * Test:
 * # Producer
 * docker exec -it kafka kafka-console-producer \
 *   --topic source-topic --bootstrap-server localhost:9092
 * 
 * # Consumer
 * docker exec -it kafka kafka-console-consumer \
 *   --topic sink-topic --bootstrap-server localhost:9092 \
 *   --from-beginning
 */
public class KafkaToKafka {

    private static final ObjectMapper objectMapper = new ObjectMapper();

    public static void main(String[] args) throws Exception {
        // 1. Create execution environment
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        // 2. Configure Kafka source
        KafkaSource<String> source = KafkaSource.<String>builder()
            .setBootstrapServers("kafka:29092")
            .setTopics("source-topic")
            .setGroupId("flink-pipeline-group")
            .setStartingOffsets(OffsetsInitializer.earliest())
            .setValueOnlyDeserializer(new SimpleStringSchema())
            .build();

        // 3. Create input stream
        DataStream<String> inputStream = env.fromSource(
            source,
            WatermarkStrategy.noWatermarks(),
            "Kafka Source"
        );

        // 4. Transform the data
        DataStream<String> processedStream = inputStream
            .map(message -> {
                try {
                    // Try to parse as JSON
                    JsonNode jsonNode = objectMapper.readTree(message);
                    ObjectNode outputNode = objectMapper.createObjectNode();
                    
                    // Copy original fields
                    outputNode.setAll((ObjectNode) jsonNode);
                    
                    // Add processing metadata
                    outputNode.put("processed", true);
                    outputNode.put("processedAt", System.currentTimeMillis());
                    outputNode.put("uppercase", message.toUpperCase());
                    
                    return objectMapper.writeValueAsString(outputNode);
                } catch (Exception e) {
                    // If not JSON, create a simple JSON structure
                    ObjectNode outputNode = objectMapper.createObjectNode();
                    outputNode.put("raw_message", message);
                    outputNode.put("processed", true);
                    outputNode.put("processedAt", System.currentTimeMillis());
                    outputNode.put("uppercase", message.toUpperCase());
                    
                    return objectMapper.writeValueAsString(outputNode);
                }
            })
            .name("Transform");

        // Also print for debugging
        processedStream.print("Processed");

        // 5. Configure Kafka sink
        KafkaSink<String> sink = KafkaSink.<String>builder()
            .setBootstrapServers("kafka:29092")
            .setRecordSerializer(
                KafkaRecordSerializationSchema.builder()
                    .setTopic("sink-topic")
                    .setValueSerializationSchema(new SimpleStringSchema())
                    .build()
            )
            .setDeliveryGuarantee(DeliveryGuarantee.AT_LEAST_ONCE)
            .build();

        // 6. Write to Kafka
        processedStream.sinkTo(sink).name("Kafka Sink");

        // 7. Execute the job
        env.execute("Kafka to Kafka Pipeline");
    }
}

File: stream-processing/flink/examples/py/kafka_to_kafka.py
"""
Apache Flink - Kafka to Kafka Pipeline
=======================================
Complete pipeline that reads from a Kafka topic, transforms the data,
and writes to another Kafka topic.

Prerequisites:
- Kafka running: cd ../kafka && docker-compose up -d
- Flink running: docker-compose up -d
- Topics 'input-topic' and 'output-topic' created
- apache-flink and apache-flink-libraries installed
"""

from pyflink.datastream import StreamExecutionEnvironment
from pyflink.datastream.connectors.kafka import (
    KafkaSource, 
    KafkaOffsetsInitializer,
    KafkaSink,
    KafkaRecordSerializationSchema
)
from pyflink.common.serialization import SimpleStringSchema
from pyflink.common.typeinfo import Types
import json


def create_kafka_pipeline():
    """
    Pipeline that reads from 'input-topic', processes, and writes to 'output-topic'.
    """
    # Create execution environment
    env = StreamExecutionEnvironment.get_execution_environment()
    env.set_parallelism(1)
    
    # Configure Kafka source (input)
    kafka_source = KafkaSource.builder() \
        .set_bootstrap_servers("kafka:29092") \
        .set_topics("input-topic") \
        .set_group_id("flink-pipeline-group") \
        .set_starting_offsets(KafkaOffsetsInitializer.earliest()) \
        .set_value_only_deserializer(SimpleStringSchema()) \
        .build()
    
    # Create input stream
    input_stream = env.from_source(
        kafka_source,
        watermark_strategy=None,
        source_name="Kafka Input"
    )
    
    # Process data
    def process_message(msg):
        """
        Example transformation: converts to JSON and adds timestamp
        """
        try:
            data = json.loads(msg)
        except:
            data = {"raw_message": msg}
        
        data["processed"] = True
        data["uppercase"] = msg.upper()
        
        return json.dumps(data)
    
    processed_stream = input_stream.map(
        process_message, 
        output_type=Types.STRING()
    )
    
    # Configure Kafka sink (output)
    kafka_sink = KafkaSink.builder() \
        .set_bootstrap_servers("kafka:29092") \
        .set_record_serializer(
            KafkaRecordSerializationSchema.builder()
                .set_topic("output-topic")
                .set_value_serialization_schema(SimpleStringSchema())
                .build()
        ) \
        .build()
    
    # Write to Kafka
    processed_stream.sink_to(kafka_sink)
    
    # Also print for debugging
    processed_stream.print()
    
    # Execute the job
    env.execute("Kafka to Kafka Pipeline")


def setup_instructions():
    """
    Prints setup instructions.
    """
    print("=" * 60)
    print("Apache Flink - Kafka to Kafka Pipeline")
    print("=" * 60)
    print("\n1. Make sure Kafka is running:")
    print("   cd ../kafka && docker-compose up -d")
    
    print("\n2. Create the required topics:")
    print("   docker exec -it kafka kafka-topics --create \\")
    print("     --topic input-topic --bootstrap-server localhost:9092 \\")
    print("     --partitions 1 --replication-factor 1")
    print()
    print("   docker exec -it kafka kafka-topics --create \\")
    print("     --topic output-topic --bootstrap-server localhost:9092 \\")
    print("     --partitions 1 --replication-factor 1")
    
    print("\n3. Produce test messages to input-topic:")
    print("   docker exec -it kafka kafka-console-producer \\")
    print("     --topic input-topic --bootstrap-server localhost:9092")
    
    print("\n4. Consume messages from output-topic (in another terminal):")
    print("   docker exec -it kafka kafka-console-consumer \\")
    print("     --topic output-topic --bootstrap-server localhost:9092 \\")
    print("     --from-beginning")
    
    print("\n5. Run this script:")
    print("   python kafka_to_kafka.py")
    print("=" * 60)
    print()


if __name__ == '__main__':
    setup_instructions()
    
    response = input("Do you want to start the pipeline? (y/n): ")
    if response.lower() == 'y':
        print("\nStarting Kafka to Kafka pipeline...")
        create_kafka_pipeline()
    else:
        print("Pipeline not started.")
